// eslint-plugin-svelte/lib/rules/experimental-require-strict-events.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("../utils");
const ast_utils_1 = require("../utils/ast-utils");
exports.default = (0, utils_1.createRule)("experimental-require-strict-events", {
    meta: {
        docs: {
            description: "require the strictEvents attribute on `<script>` tags",
            category: "Experimental",
            recommended: false,
        },
        schema: [],
        messages: {
            missingStrictEvents: `The component must have the strictEvents attribute on its <script> tag or it must define the $$Events interface.`,
        },
        type: "suggestion",
    },
    create(context) {
        let isTs = false;
        let hasAttribute = false;
        let hasInterface = false;
        let scriptNode;
        return {
            SvelteScriptElement(node) {
                const lang = (0, ast_utils_1.getLangValue)(node)?.toLowerCase();
                isTs = lang === "ts" || lang === "typescript";
                hasAttribute = (0, ast_utils_1.findAttribute)(node, "strictEvents") !== null;
                scriptNode = node;
            },
            TSInterfaceDeclaration(node) {
                if (node.id.name === "$$Events") {
                    hasInterface = true;
                }
            },
            "Program:exit"() {
                if (isTs && !hasAttribute && !hasInterface) {
                    context.report({
                        node: scriptNode,
                        messageId: "missingStrictEvents",
                    });
                }
            },
        };
    },
});
